### PR TITLE
Add Raw Temperature Dimensions Metrics to the Summaries

### PR DESCRIPTION
**Pull request by aditjind**
*Issue #, if available:* Add Raw Metrics to the Summaries #222 

*Description of changes:* Adding raw metrics based on which temperature profile is calculated to the Shard, Node, Zone and Cluster summaries. Fixing bugs for the exception in operate in case the metric table is not found in the window. 

*Tests:* UTs addition in progress, Manual Testing done. Current Build is failing due to addition of UT pending.   

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java [hunk 1]
```diff
@@ -72,11 +73,17 @@
      */
     private TemperatureVector.NormalizedValue meanTemperature;
 
+    /**
+     * This is the average value used over shards.
+     */
+
```
Comment (rguo-aws) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java:79]: remove this line ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java [hunk 2]
```diff
@@ -243,7 +261,11 @@ public static ClusterDimensionalSummary build(Record record, DSLContext context)
 
         ClusterDimensionalSummary summary =
                 new ClusterDimensionalSummary(TemperatureDimension.valueOf(dimensionName));
-        summary.setTotalUsage(total);
+        if (numNodes <= 0) {
+            throw new IllegalArgumentException("Number of elasticsearch nodes stored are zero.");
```
Comment (rguo-aws) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java:265]: add IllegalArgumentException to this build method declaration ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/CpuUtilDimensionTemperatureRca.java [hunk 3]
```diff
@@ -73,9 +73,15 @@ public DimensionalTemperatureFlowUnit operate() {
...
         LOG.info("CPU Utilization temperature calculated: {}",
                 cpuUtilTemperatureFlowUnit.getNodeDimensionProfile());
-        ResourceContext context = (cpuUtilTemperatureFlowUnit.getNodeDimensionProfile().getMeanTemperature()
-                .isGreaterThan(THRESHOLD_NORMALIZED_VAL_FOR_HEAT_ZONE_ASSIGNMENT)) ? new ResourceContext(Resources.State.UNHEALTHY) :
-                new ResourceContext(Resources.State.HEALTHY);
+        ResourceContext context;
+        try {
+            context = (cpuUtilTemperatureFlowUnit.getNodeDimensionProfile().getMeanTemperature()
+                    .isGreaterThan(THRESHOLD_NORMALIZED_VAL_FOR_HEAT_ZONE_ASSIGNMENT)) ? new ResourceContext(Resources.State.UNHEALTHY) :
+                    new ResourceContext(Resources.State.HEALTHY);
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/CpuUtilDimensionTemperatureRca.java:80]: We should remove this as the health based on temperature is not inferred at this level. It happens at a higher level component - pyrometer.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/HeapAllocRateTemperatureRca.java [hunk 4]
```diff
@@ -63,9 +63,15 @@ public DimensionalTemperatureFlowUnit operate() {
               THRESHOLD_NORMALIZED_VAL_FOR_HEAT_ZONE_ASSIGNMENT);
       LOG.info("Heap allocation rate temperature calculated: {}",
               heapAllocRateTemperatureFlowUnit.getNodeDimensionProfile());
-      ResourceContext context = (heapAllocRateTemperatureFlowUnit.getNodeDimensionProfile().getMeanTemperature()
-              .isGreaterThan(THRESHOLD_NORMALIZED_VAL_FOR_HEAT_ZONE_ASSIGNMENT)) ? new ResourceContext(Resources.State.UNHEALTHY) :
-              new ResourceContext(Resources.State.HEALTHY);
+      ResourceContext context;
+      try {
+          context = (heapAllocRateTemperatureFlowUnit.getNodeDimensionProfile().getMeanTemperature()
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/HeapAllocRateTemperatureRca.java:68]: same

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ResourceHeatMapGraphTest.java [hunk 5]
```diff
@@ -39,12 +39,15 @@
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.temperature.ClusterTemperatureSummary;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.temperature.CompactClusterLevelNodeSummary;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.temperature.NodeLevelDimensionalSummary;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.api.summaries.temperature.ShardProfileSummary;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.ConnectedComponent;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.Queryable;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.RcaConf;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.Stats;
 import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.temperature.HeatZoneAssigner;
+import com.amazon.opendistro.elasticsearch.performanceanalyzer.rca.framework.core.temperature.RawMetricsVector;
```
Comment (yojs) [inline src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ResourceHeatMapGraphTest.java:48]: We metric values are absolute metric value counterparts for the normalized values. Can't we store the metric values as an additional field in the NormalizedValue class ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/CpuUtilDimensionTemperatureRca.java (hunk 3 above)
Comment (aditjind) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/CpuUtilDimensionTemperatureRca.java:80]: Done

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/HeapAllocRateTemperatureRca.java (hunk 4 above)
Comment (aditjind) [reply src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/rca/temperature/dimension/HeapAllocRateTemperatureRca.java:68]: Done

File: src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ResourceHeatMapGraphTest.java (hunk 5 above)
Comment (aditjind) [reply src/test/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/store/ResourceHeatMapGraphTest.java:48]: Have refactored the code to take this into affect.

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java [hunk 6]
```diff
@@ -70,13 +71,19 @@
...
+     */
+
+    private double avgMetricValueUsedOverNodes;
 
     /**
      * meanTemperature is a normalized value. The total tells us if this is something that one
      * should be concerned about.
      */
-    private double totalUsage;
+    private double totalMetricValueUsedOverCluster;
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java:86]: Does the substring `OverCluster` and `OverNodes` signify something different ? If not, then we can remove them as the name of the class tells us that this is something over all the nodes? Thoughts ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java [hunk 7]
```diff
@@ -70,13 +71,19 @@
...
      * This is the mean temperature for this dimension over all the nodes in the cluster.
      */
-    private TemperatureVector.NormalizedValue meanTemperature;
+    private TemperatureVector.VectorValues meanTemperature;
+
+    /**
+     * This is the average value used over shards.
+     */
+
+    private double avgMetricValueUsedOverNodes;
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java:80]: The `VectorValues` class now contains the raw value and the normalized value. So, shouldn't this be already covered in the rawValue part ? Do we really need a different variable for this ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java [hunk 8]
```diff
@@ -98,12 +105,16 @@ public ClusterDimensionalSummary(TemperatureDimension profileForDimension) {
...
     }
 
-    public void setTotalUsage(double totalUsage) {
-        this.totalUsage = totalUsage;
+    public void setAvgMetricValueUsedOverNodes(double avgMetricValueUsedOverNodes) {
+        this.avgMetricValueUsedOverNodes = avgMetricValueUsedOverNodes;
+    }
+
+    public void setTotalMetricsValueUsed(double totalMetricValueUsedOverCluster) {
+        this.totalMetricValueUsedOverCluster = totalMetricValueUsedOverCluster;
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java:117]: Average and total are over all shards in the cluster, right ? We should add that in the name to make it clear ?

File: src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java [hunk 9]
```diff
@@ -173,6 +188,7 @@ public String getTableName() {
         List<Field<?>> schema = new ArrayList<>();
         schema.add(DSL.field(DSL.name(DIM_KEY), String.class));
         schema.add(DSL.field(DSL.name(MEAN_KEY), Short.class));
+        schema.add(DSL.field(DSL.name(AVG_NODE_KEY), Short.class));
```
Comment (yojs) [inline src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/rca/framework/api/summaries/temperature/ClusterDimensionalSummary.java:191]: Let's use the word either mean or avg and not both.

**Pull request closed without merging**